File: software/Servo.py
import RPi.GPIO as GPIO
import time

class Servo:
    def __init__(self,gpioPin,servo_range,desired_servo_range,min_pulse_micro_seconds,max_pulse_micro_seconds,frequency):
        self.currentAngle = 0 
        self.frequency = frequency
        self.angle_max_range = desired_servo_range if desired_servo_range == None else servo_range
        self.min_pulse_micro_seconds = min_pulse_micro_seconds
        self.max_pulse_micro_seconds = max_pulse_micro_seconds
        self.gpioPin = gpioPin
        
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(gpioPin,GPIO.OUT)
        self.pwm = GPIO.PWM(gpioPin,frequency)
        self.pwm.start(self.getDutyCycle(self.currentAngle))

    def __del__(self):
        self.pwm.stop()
        GPIO.cleanup()

    def getDutyCycle(self,degrees):
        servo_pulse_micro_seconds = 10**6 / self.frequency # (10^6 microseconds per second) / (# cycles per second) =  # seconds / cycle 
        pulse_range_micro_seconds = self.max_pulse_micro_seconds - self.min_pulse_micro_seconds
        percent_angle = degrees / self.angle_max_range
        target_pulse_micro_seconds = self.min_pulse_micro_seconds + (pulse_range_micro_seconds * percent_angle)
        dutyCycle = (target_pulse_micro_seconds / servo_pulse_micro_seconds) * 100
        return dutyCycle

    def turn(self,degrees):
        if degrees + self.currentAngle > self.angle_max_range:
            self.currentAngle = self.angle_max_range
        elif degrees + self.currentAngle < 0:
            self.currentAngle = 0
        else:
            self.currentAngle += degrees
        
        newDutyCycle = self.getDutyCycle(self.currentAngle)
        self.pwm.ChangeDutyCycle(newDutyCycle)

if __name__=='__main__':
    servo1 = Servo(
        gpioPin=17,
        servo_range=270,
        desired_servo_range=80,
        min_pulse_micro_seconds=500,
        max_pulse_micro_seconds=2500,
        frequency=50
    )

    servo2 = Servo(
        gpioPin=24,
        servo_range=270,
        desired_servo_range=180,
        min_pulse_micro_seconds=500,
        max_pulse_micro_seconds=2500,
        frequency=50
    )
    
    servo1.turn(40)
    servo2.turn(90)
    time.sleep(1)
    servo1.turn(40)
    servo2.turn(90)
    time.sleep(1)
    servo1.turn(-90)
    servo2.turn(-180)
    time.sleep(1)

File: rust-rewrite/src/servo.rs
use getset::Getters;
use rppal::pwm;
use std::cmp;

#[derive(Getters)]
pub struct Servo {
    pwm: pwm::Pwm,
    #[getset(get = "pub")]
    current_angle: f64,
    #[getset(get = "pub")]
    full_servo_range: u16, 
    #[getset(get = "pub")]
    desired_servo_range: u16, 
    #[getset(get = "pub")]
    min_pulse_microseconds: u32, 
    #[getset(get = "pub")]
    max_pulse_microseconds: u32, 
    #[getset(get = "pub")]
    frequency: u16,
}

impl Servo {
    pub fn new(channel: pwm::Channel, desired_servo_range: u16, full_servo_range: u16, min_pulse_microseconds: u32, max_pulse_microseconds: u32, frequency: u16) -> Self {
        let pwm: Result<pwm::Pwm, pwm::Error> = pwm::Pwm::with_frequency(channel, frequency, 0, pwm::Polarity::Normal, true);
        return Servo {
            pwm: pwm,
            current_angle: 0,
            full_servo_range: full_servo_range,
            desired_servo_range: desired_servo_range,
            min_pulse_microseconds: min_pulse_microseconds,
            max_pulse_microseconds: max_pulse_microseconds,
            frequency: frequency
        };
    }

    fn get_duty_cycle(&self, angle: f64) -> f64 {
        let servo_pulse_microseconds = 1_000_000.0 / self.frequency as f64;
        let pulse_range_microseconds = (self.max_pulse_microseconds - self.min_pulse_microseconds) as f64;
        let percent_angle = (angle / self.full_servo_range) as f64;
        let target_pulse_microseconds = self.min_pulse_microseconds as f64 + (pulse_range_microseconds * percent_angle);
        let duty_cycle = (target_pulse_microseconds / servo_pulse_microseconds) * 100.0;
        return duty_cycle;
    }

    pub fn turn(&self, angle: f64) {
        if(angle + self.current_angle > self.desired_servo_range) {
            self.current_angle = self.desired_servo_range;
        } else if() {
            self.current_angle = 0;
        } else {
            self.current_angle += angle;
        }

        let new_duty_cycle = self.get_duty_cycle(self.current_angle);
        self.pwm.set_duty_cycle(new_duty_cycle)
    }
}

pub fn main() -> Result<(), Box<dyn Error>> {
    let x_servo = Servo::new(0,270,270,500,2500,50);
    let y_servo = Servo::new(1,270,270,500,2500,50);
    x_servo.turn(135);
    y_servo.turn(135);
    x_servo.turn(135);
    y_servo.turn(135);
    x_servo.turn(-270);
    y_servo.turn(-270);
    Ok(())
}